README.md
node-speaker
============
### Output [PCM audio][pcm] data to the speakers
[![Build Status](https://secure.travis-ci.org/TooTallNate/node-speaker.svg)](https://travis-ci.org/TooTallNate/node-speaker)
[![Build Status](https://ci.appveyor.com/api/projects/status/wix7wml3v55670kw?svg=true)](https://ci.appveyor.com/project/TooTallNate/node-speaker)


A Writable stream instance that accepts [PCM audio][pcm] data and outputs it
to the speakers. The output is backed by `mpg123`'s audio output modules, which
in turn use any number of audio backends commonly found on Operating Systems
these days.


Installation
------------

Simply compile and install `node-speaker` using `npm`:

``` bash
$ npm install speaker
```

On Debian/Ubuntu, the [ALSA][alsa] backend is selected by default, so be sure
to have the `alsa.h` header file in place:

``` bash
$ sudo apt-get install libasound2-dev
```


Example
-------

Here's an example of piping `stdin` to the speaker, which should be 2 channel,
16-bit audio at 44,100 samples per second (a.k.a CD quality audio).

``` javascript
var Speaker = require('speaker');

// Create the Speaker instance
var speaker = new Speaker({
  channels: 2,          // 2 channels
  bitDepth: 16,         // 16-bit samples
  sampleRate: 44100     // 44,100 Hz sample rate
});

// PCM data from stdin gets piped into the speaker
process.stdin.pipe(speaker);
```


API
---

`require('speaker')` directly returns the `Speaker` constructor. It is the only
interface exported by `node-speaker`.

### new Speaker([ format ]) -> Speaker instance;

Creates a new `Speaker` instance, which is a writable stream that you can pipe
PCM audio data to. The optional `format` object may contain any of the `Writable`
base class options, as well as any of these PCM formatting options:

  * `channels` - The number of audio channels. PCM data must be interleaved. Defaults to `2`.
  * `bitDepth` - The number of bits per sample. Defaults to `16` (16-bit).
  * `sampleRate` - The number of samples per second per channel. Defaults to `44100`.
  * `signed` - Boolean specifying if the samples are signed or unsigned. Defaults to `true` when bit depth is 8-bit, `false` otherwise.
  * `float` - Boolean specifying if the samples are floating-point values. Defaults to `false`.
  * `samplesPerFrame` - The number of samples to send to the audio backend at a time. You likely don't need to mess with this value. Defaults to `1024`.

#### "open" event

Fired when the backend `open()` call has completed. This happens once the first
`write()` call happens on the speaker instance.

#### "flush" event

Fired after the speaker instance has had `end()` called, and after the audio data
has been flushed to the speakers.

#### "close" event

Fired after the "flush" event, after the backend `close()` call has completed.
This speaker instance is essentially finished after this point.


Audio Backend Selection
-----------------------

`node-speaker` is backed by `mpg123`'s "output modules", which in turn use one of
many popular audio backends like ALSA, OSS, SDL, and lots more. The default
backends for each operating system are described in the table below:

| **Operating System** | **Audio Backend** | **Description**
|:---------------------|:------------------|:----------------------------------
| Linux                | `alsa`            | Output audio using [Advanced Linux Sound Architecture (ALSA)][alsa].
| Mac OS X             | `coreaudio`       | Output audio using Mac OS X's CoreAudio.
| Windows              | `win32`           | Audio output for Windows (winmm).
| Solaris              | `sun`             | Audio output for Sun Audio.

To manually override the default backend, pass the `--mpg123-backend` switch to
`npm`/`node-gyp`:

``` bash
$ npm install speaker --mpg123-backend=openal
```

[pcm]: http://en.wikipedia.org/wiki/Pulse-code_modulation
[alsa]: http://www.alsa-project.org/


History.md

0.3.0 / 2016-04-09
==================

  * Update to latest version win32.c
  * README: use SVG for appveyor badge

0.2.6 / 2015-09-24
==================

  * more travis/appveyor testing
  * src: modification so it works with nan@2 (#64, @santigimeno)
  * fix compile with io.js 2.1.0 on Linux (#56, @bradjc)
  * made it compile on freebsd (#55, @antxxxx)

0.2.5 / 2015-04-15
==================

  * package: update "bindings" to v1.2.1
  * package: update "nan" to v1.7.0
  * appveyor: test node v0.12, don't test v0.11
  * travis: test node v0.12, don't test v0.11

0.2.4 / 2015-02-20
==================

  * update "nan" to v1.6.2
  * update binding.cc for node v0.12 / iojs support (#47, @mrinalvirnave)

0.2.3 / 2015-02-20
==================

  * link pulse-simple when using mpg123 "pulse" backend (#48, @ReneHollander)

0.2.2 / 2015-01-13
==================

  * example: make sine.js produce 220hz, not 440hz (#38, @jproulx)
  * package: update "mocha" to v2.1.0
  * package: allow any "debug" v2
  * package: add "bugs" and "homepage" fields
  * travis: remove unused "matrix" field

0.2.1 / 2014-07-07
==================

  * src: call open() and close() on the dummy audio_output_t instance (#36)
  * README: document the `float` format option

0.2.0 / 2014-06-22
==================

  * gitignore: ignore root-level dev files
  * index: pass the `format` directly to the native open() binding
  * index: ensure a valid and supported "format" is given to Speaker instance
  * test: add quotes to test names
  * index: add `getFormat()` and `isSupported()` functions
  * binding: export the `MPG123_ENC_*` constants
  * binding: export the result of `get_formats()`
  * mpg123: add 64-bit float playback support for CoreAudio backend
  * index: use %o formatter some more
  * index: default `float` to 32-bit `bitDepth`

0.1.3 / 2014-06-15
==================

  * index: do not call `flush()` binding when the stream finishes "naturally"
  * index: use %o formatting from debug v1
  * package: update "nan" to v1.2.0
  * package: update "debug" to v1.0.0

0.1.2 / 2014-06-02
==================

  * package: update "nan" to v1.1.2
  * package: update "mocha" dev dependency

0.1.1 / 2014-05-27
==================

  * binding: update to nan v1.1.0 API, fixes node v0.11.13+
  * add appveyor.yml file for Windows testing
  * README: add appveyor build badge
  * README: use svg for travis badge
  * travis: don't test node v0.9.x
  * index: make _format() bind to the speaker instance
  * don't leave event listeners behind (#22, @LinusU)

0.1.0 / 2014-04-17
==================

  * index: abort write() call if `_close` is set (#28, #29)
  * package: tighten up the dependencies' versions
  * index: add a debug() call
  * index: emit "close" after setting `_closed`
  * index: use the "readable-stream" copy of Writable
  * package: pin "readable-stream" to any v1.0.x
  * examples: fix "sine" emitting "end" event
  * travis: test node v0.11
  * use `rvagg/nan`
  * fix History.md note
  * fix sinewave example on 0.10 (Stream API changes) (#12, @jfmatt)

0.0.10 / 2013-05-08
===================

  * pass the "open" error to the Speaker instance. Closes #7.
  * package: add "sound" as a keyword
  * travis: test node v0.10

0.0.9 / 2013-03-06
==================

  * update for v0.9.12 Writable stream API change
  * a couple more jsdoc comments

0.0.8 / 2013-02-10
==================

  * throw an Error if non-native endianness is specified

0.0.7 / 2013-01-14
==================

  * wait for the `format` event on pipe'd Readable instances
  * default the lowWaterMark and highWaterMark to 0
  * rename _opts() to _format()
  * package: allow any "readable-stream" version
  * add a few more debug calls

0.0.6 / 2012-12-15
==================

  * add node >= v0.9.4 compat

0.0.5 / 2012-11-16
==================

  * add initial tests (uses the "dummy" output module)
  * add "float" (32-bit and 64-bit) output support
  * ensure only one "close" event

0.0.4 / 2012-11-04
==================

  * mpg123: add linux arm support
  * guard against bindings that don't have a `deinit()` function

0.0.3 / 2012-11-03
==================

  * a two examples to the "examples" dir
  * emit an "open" event
  * emit a "close" event
  * emit a "flush" event
  * properly support the "pipe" event
  * mpg123: fix a CoreAudio backend compilation warning
  * add a timeout after the flush call to ensure the backend has time to play

0.0.2 / 2012-10-25
==================

  * support for Windows
  * support for Linux
  * support for Solaris
  * call `flush()` and `close()` at the end of the stream

0.0.1 / 2012-10-24
==================

  * Initial release


index.js

/**
 * Module dependencies.
 */

var os = require('os');
var debug = require('debug')('speaker');
var binding = require('bindings')('binding');
var inherits = require('util').inherits;
var Writable = require('readable-stream/writable');

// determine the native host endianness, the only supported playback endianness
var endianness = 'function' == os.endianness ?
                 os.endianness() :
                 'LE'; // assume little-endian for older versions of node.js

/**
 * Module exports.
 */

exports = module.exports = Speaker;

/**
 * Export information about the `mpg123_module_t` being used.
 */

exports.api_version = binding.api_version;
exports.description = binding.description;
exports.module_name = binding.name;

/**
 * Returns the `MPG123_ENC_*` constant that corresponds to the given "format"
 * object, or `null` if the format is invalid.
 *
 * @param {Object} format - format object with `channels`, `sampleRate`, `bitDepth`, etc.
 * @return {Number} MPG123_ENC_* constant, or `null`
 * @api public
 */

exports.getFormat = function getFormat (format) {
  var f = null;
  if (format.bitDepth == 32 && format.float && format.signed) {
    f = binding.MPG123_ENC_FLOAT_32;
  } else if (format.bitDepth == 64 && format.float && format.signed) {
    f = binding.MPG123_ENC_FLOAT_64;
  } else if (format.bitDepth == 8 && format.signed) {
    f = binding.MPG123_ENC_SIGNED_8;
  } else if (format.bitDepth == 8 && !format.signed) {
    f = binding.MPG123_ENC_UNSIGNED_8;
  } else if (format.bitDepth == 16 && format.signed) {
    f = binding.MPG123_ENC_SIGNED_16;
  } else if (format.bitDepth == 16 && !format.signed) {
    f = binding.MPG123_ENC_UNSIGNED_16;
  } else if (format.bitDepth == 24 && format.signed) {
    f = binding.MPG123_ENC_SIGNED_24;
  } else if (format.bitDepth == 24 && !format.signed) {
    f = binding.MPG123_ENC_UNSIGNED_24;
  } else if (format.bitDepth == 32 && format.signed) {
    f = binding.MPG123_ENC_SIGNED_32;
  } else if (format.bitDepth == 32 && !format.signed) {
    f = binding.MPG123_ENC_UNSIGNED_32;
  }
  return f;
}

/**
 * Returns `true` if the given "format" is playable via the "output module"
 * that was selected during compilation, or `false` if not playable.
 *
 * @param {Number} format - MPG123_ENC_* format constant
 * @return {Boolean} true if the format is playable, false otherwise
 * @api public
 */

exports.isSupported = function isSupported (format) {
  if ('number' !== typeof format) format = exports.getFormat(format);
  return (binding.formats & format) === format;
}

/**
 * The `Speaker` class accepts raw PCM data written to it, and then sends that data
 * to the default output device of the OS.
 *
 * @param {Object} opts options object
 * @api public
 */

function Speaker (opts) {
  if (!(this instanceof Speaker)) return new Speaker(opts);

  // default lwm and hwm to 0
  if (!opts) opts = {};
  if (null == opts.lowWaterMark) opts.lowWaterMark = 0;
  if (null == opts.highWaterMark) opts.highWaterMark = 0;

  Writable.call(this, opts);

  // chunks are sent over to the backend in "samplesPerFrame * blockAlign" size.
  // this is necessary because if we send too big of chunks at once, then there
  // won't be any data ready when the audio callback comes (experienced with the
  // CoreAudio backend)
  this.samplesPerFrame = 1024;

  // the `audio_output_t` struct pointer Buffer instance
  this.audio_handle = null;

  // flipped after close() is called, no write() calls allowed after
  this._closed = false;

  // set PCM format
  this._format(opts);

  // bind event listeners
  this._format = this._format.bind(this);
  this.on('finish', this._flush);
  this.on('pipe', this._pipe);
  this.on('unpipe', this._unpipe);
}
inherits(Speaker, Writable);

/**
 * Calls the audio backend's `open()` function, and then emits an "open" event.
 *
 * @api private
 */

Speaker.prototype._open = function () {
  debug('open()');
  if (this.audio_handle) {
    throw new Error('_open() called more than once!');
  }
  // set default options, if not set
  if (null == this.channels) {
    debug('setting default %o: %o', 'channels', 2);
    this.channels = 2;
  }
  if (null == this.bitDepth) {
    var depth = this.float ? 32 : 16;
    debug('setting default %o: %o', 'bitDepth', depth);
    this.bitDepth = depth;
  }
  if (null == this.sampleRate) {
    debug('setting default %o: %o', 'sampleRate', 44100);
    this.sampleRate = 44100;
  }
  if (null == this.signed) {
    debug('setting default %o: %o', 'signed', this.bitDepth != 8);
    this.signed = this.bitDepth != 8;
  }

  var format = exports.getFormat(this);
  if (null == format) {
    throw new Error('invalid PCM format specified');
  }

  if (!exports.isSupported(format)) {
    throw new Error('specified PCM format is not supported by "' + binding.name + '" backend');
  }

  // calculate the "block align"
  this.blockAlign = this.bitDepth / 8 * this.channels;

  // initialize the audio handle
  // TODO: open async?
  this.audio_handle = new Buffer(binding.sizeof_audio_output_t);
  var r = binding.open(this.audio_handle, this.channels, this.sampleRate, format);
  if (0 !== r) {
    throw new Error('open() failed: ' + r);
  }

  this.emit('open');
  return this.audio_handle;
};

/**
 * Set given PCM formatting options. Called during instantiation on the passed in
 * options object, on the stream given to the "pipe" event, and a final time if
 * that stream emits a "format" event.
 *
 * @param {Object} opts
 * @api private
 */

Speaker.prototype._format = function (opts) {
  debug('format(object keys = %o)', Object.keys(opts));
  if (null != opts.channels) {
    debug('setting %o: %o', 'channels', opts.channels);
    this.channels = opts.channels;
  }
  if (null != opts.bitDepth) {
    debug('setting %o: %o', "bitDepth", opts.bitDepth);
    this.bitDepth = opts.bitDepth;
  }
  if (null != opts.sampleRate) {
    debug('setting %o: %o', "sampleRate", opts.sampleRate);
    this.sampleRate = opts.sampleRate;
  }
  if (null != opts.float) {
    debug('setting %o: %o', "float", opts.float);
    this.float = opts.float;
  }
  if (null != opts.signed) {
    debug('setting %o: %o', "signed", opts.signed);
    this.signed = opts.signed;
  }
  if (null != opts.samplesPerFrame) {
    debug('setting %o: %o', "samplesPerFrame", opts.samplesPerFrame);
    this.samplesPerFrame = opts.samplesPerFrame;
  }
  if (null == opts.endianness || endianness == opts.endianness) {
    // no "endianness" specified or explicit native endianness
    this.endianness = endianness;
  } else {
    // only native endianness is supported...
    this.emit('error', new Error('only native endianness ("' + endianness + '") is supported, got "' + opts.endianness + '"'));
  }
};

/**
 * `_write()` callback for the Writable base class.
 *
 * @param {Buffer} chunk
 * @param {String} encoding
 * @param {Function} done
 * @api private
 */

Speaker.prototype._write = function (chunk, encoding, done) {
  debug('_write() (%o bytes)', chunk.length);

  if (this._closed) {
    // close() has already been called. this should not be called
    return done(new Error('write() call after close() call'));
  }
  var b;
  var self = this;
  var left = chunk;
  var handle = this.audio_handle;
  if (!handle) {
    // this is the first time write() is being called; need to _open()
    try {
      handle = this._open();
    } catch (e) {
      return done(e);
    }
  }
  var chunkSize = this.blockAlign * this.samplesPerFrame;

  function write () {
    if (self._closed) {
      debug('aborting remainder of write() call (%o bytes), since speaker is `_closed`', left.length);
      return done();
    }
    b = left;
    if (b.length > chunkSize) {
      var t = b;
      b = t.slice(0, chunkSize);
      left = t.slice(chunkSize);
    } else {
      left = null;
    }
    debug('writing %o byte chunk', b.length);
    binding.write(handle, b, b.length, onwrite);
  }

  function onwrite (r) {
    debug('wrote %o bytes', r);
    if (r != b.length) {
      done(new Error('write() failed: ' + r));
    } else if (left) {
      debug('still %o bytes left in this chunk', left.length);
      write();
    } else {
      debug('done with this chunk');
      done();
    }
  }

  write();
};

/**
 * Called when this stream is pipe()d to from another readable stream.
 * If the "sampleRate", "channels", "bitDepth", and "signed" properties are
 * set, then they will be used over the currently set values.
 *
 * @api private
 */

Speaker.prototype._pipe = function (source) {
  debug('_pipe()');
  this._format(source);
  source.once('format', this._format);
};

/**
 * Called when this stream is pipe()d to from another readable stream.
 * If the "sampleRate", "channels", "bitDepth", and "signed" properties are
 * set, then they will be used over the currently set values.
 *
 * @api private
 */

Speaker.prototype._unpipe = function (source) {
  debug('_unpipe()');
  source.removeListener('format', this._format);
};

/**
 * Emits a "flush" event and then calls the `.close()` function on
 * this Speaker instance.
 *
 * @api private
 */

Speaker.prototype._flush = function () {
  debug('_flush()');
  this.emit('flush');
  this.close(false);
};

/**
 * Closes the audio backend. Normally this function will be called automatically
 * after the audio backend has finished playing the audio buffer through the
 * speakers.
 *
 * @param {Boolean} flush - if `false`, then don't call the `flush()` native binding call. Defaults to `true`.
 * @api public
 */

Speaker.prototype.close = function (flush) {
  debug('close(%o)', flush);
  if (this._closed) return debug('already closed...');

  if (this.audio_handle) {
    if (false !== flush) {
      // TODO: async most likely…
      debug('invoking flush() native binding');
      binding.flush(this.audio_handle);
    }

    // TODO: async maybe?
    debug('invoking close() native binding');
    binding.close(this.audio_handle);
    this.audio_handle = null;
  } else {
    debug('not invoking flush() or close() bindings since no `audio_handle`');
  }

  this._closed = true;
  this.emit('close');
};
